### DISCUSSION TITLE
Rotate map marker

### DISCUSSION THREAD

**Discussion by AFA141** [mapsforge/mapsforge#1329] [Q&A] [answered]
Hello, i want to ask what is the best way to rotate map marker? I tried extending JPanel by Layer class, then overriding paintComponent method in Marker class but i cant invoke paintComponent by calling repaint method on marker instance...

**Comment by devemux86** [comment]
Please specify in your questions the platform: Android or Desktop.

You can extend the `Marker` class, override the `draw` method, etc.
to perform any custom rendering.

**Reply by AFA141** [reply]
Desktop, I tried with this:

Marker override this method, and Layer extends JPanel
@Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D gx = (Graphics2D) g;
        gx.rotate(Math.toRadians(90), this.bitmap.getWidth() / 2, this.bitmap.getHeight() / 2);     
    }

and marker.repaint();

**Reply by AFA141** [reply]
I want to rotate by angle like in this example
gx.rotate(Math.toRadians(90), this.bitmap.getWidth() / 2, this.bitmap.getHeight() / 2);

**Comment by devemux86** [comment, marked as answer]
- Override `Marker.draw`, get `AwtGraphicFactory.getGraphics(Canvas)` and use directly the `Graphics2D`
- Or try the `canvas.drawBitmap(Bitmap, Matrix)`
- Or rotate the bitmap externally and use `Marker.setBitmap` to update it

You need to study the [Mapsforge graphics API](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-awt/src/main/java/org/mapsforge/map/awt/graphics/AwtCanvas.java).
(API is common on all supported platforms)

**Comment by devemux86** [comment]
See also this discussion in the old forum:
https://groups.google.com/g/mapsforge-dev/c/FD_8_iRW0Mw/m/urguOVDpL7cJ

A similar implementation could also work on the Desktop.

**Comment by AFA141** [comment]
Thanks! I finished this by overriding draw method in Marker class.

Graphics2D g = AwtGraphicFactory.getGraphics(canvas);
AffineTransform at = new AffineTransform();
at.translate(left, top);
at.rotate(Math.toRadians(90));
g.drawRenderedImage(AwtGraphicFactory.getBitmap(bitmap), at);

**Comment by devemux86** [comment]
> I finished this by overriding draw method in Marker class.

Thanks for sharing the implementation,
it can be useful to other users as well.
